Cancun updates

fselmo opened the pull request:
### What was wrong?

- Add support for type `3` blob transactions
- Refactor typed transactions into separate files for better organization
- Normalize TypedTransaction.as_dict() so that tupleized or list-based `accessList` or `blobVersionedHashes` within a dict all normalize to tuples for consistency and ease of comparison.

### Todo:

- [x] Add or update documentation related to these changes
- [x] Add entry to the [release notes](https://github.com/ethereum/eth-account/blob/main/newsfragments/README.md)
- [x] Implement a new API to wrap blob txs with the appropriate blobs, commitments, and proofs
- [x] Add tests for the wrapped `PooledTransaction` API

#### Cute Animal Picture

<img width="1277" alt="Screenshot 2024-03-19 at 09 53 18" src="https://github.com/ethereum/eth-account/assets/3532824/e45f97ca-2cad-422c-ad7e-3f631b2a6c69">

fselmo committed:
Re-organize typed transaction file structure + add blob txns:

- Re-organize typed transactions as a directory with each `_TypedTransactionImplementation`
  as its own file.
- Add support for `BlobTransaction` (type = 3).
- Add relevant tests.

Also:

- Normalize transaction dict for ``TypedTransaction.as_dict()`` so that lists or tuples are converted to tuples for consistency and ease of comparison.

pacrob reviewed (approved):
lgtm! Nice reorg

kclowes commented on eth_account/_utils/typed_transactions/base.py:
```diff
@@ -0,0 +1,78 @@
+)
+
+
+class _TypedTransactionImplementation(ABC):
```
Should `assert_valid_fields`, `from_dict` or `from_bytes` be in here at all?

fselmo replied:
I think it makes sense to have `assert_valid_fields` there but `from_dict` and `from_bytes` return instances specific to the class that they are defined within so I think the signature differences muddy the abstract method + implementation pattern.

fselmo replied:
I could see it going either way... `assert_valid_fields` isn't implemented at the `TypedTransaction` level so I don't think it's a necessity, more of a safety check. And it has a different signature for `BlobTransaction` (`has_blobs: bool = false`) that we'd have to apply across the board since we don't validate certain fields of the dictionary if blobs are also provided.

Curious on thoughts here @pacrob @kclowes

kclowes replied:
Okay, yeah, I think it makes sense to exclude them then. Thanks for the clarification!

kclowes commented on tests/core/test_typed_transactions.py:
```diff
@@ -3,9 +3,16 @@
 import pytest
 
-from eth_account._utils.typed_transactions import (
+from eth_account._utils.typed_transactions.access_list_transaction import (
```
nit: you should be able to import these from `eth_account._utils.typed_transactions` still with the `__init__.py`, I believe

fselmo committed:
Implement extended blob rlp logic for 4844 definition:

- When a blob transaction is signed, it needs to contain the blob information
  along with commitments and proofs for the consensus layer. We need to be
  able to serialize and deserialize this information.
- Use different serializers for execution layer style `TransactionPayloadBody`
  blob transactions and for `PooledTransaction` blob transactions.

kclowes commented on eth_account/_utils/typed_transactions/blob_transaction.py:
```diff
@@ -0,0 +1,242 @@
+    @classmethod
+    def from_dict(cls, dictionary: Dict[str, Any]) -> "BlobTransaction":
+        """
+        Builds a DynamicFeeTransaction from a dictionary.
```
```suggestion
        Builds a BlobTransaction from a dictionary.
```

kclowes commented on eth_account/_utils/typed_transactions/blob_transaction.py:
```diff
@@ -0,0 +1,242 @@
+
+    @classmethod
+    def from_bytes(cls, encoded_transaction: HexBytes) -> "BlobTransaction":
+        """Builds a DynamicFeeTransaction from a signed encoded transaction."""
```
```suggestion
        """Builds a BlobTransaction from a signed encoded transaction."""
```

kclowes reviewed (approved):
Looks good to me. Just left a few nits

fselmo committed:
Add more testing around blob transactions:

- Add some testing around blob transaction `blobVersionedHashes` validation
  when present in tx dict compared to the computed values from blobs.
- Add testing around the computing of commitment, proofs, and versioned hashes.
- Add testing around the blob count limits for a blob carrying transaction.

pacrob reviewed (approved):
Apart from the comment on fields to add to `_TypedTransactionImplementation`, lgtm!

fselmo committed:
UX improvements to transaction type interpretation
